foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/version_specific/MlTrainPipelineDockerMigration.java

```diff
@@ -0,0 +1,118 @@
+package com.boozallen.aissemble.upgrade.migration.version_specific;
+
+/*-
+ * #%L
+ * aiSSEMBLE::Foundation::Upgrade
+ * %%
+ * Copyright (C) 2021 Booz Allen
+ * %%
+ * This software package is licensed under the Booz Allen Public License. All Rights Reserved.
+ * #L%
+ */
+
+import com.boozallen.aissemble.upgrade.migration.AbstractPomMigration;
+import static org.technologybrewery.baton.util.pom.LocationAwareMavenReader.END;
+import static org.apache.commons.lang3.StringUtils.repeat;
+
+import org.apache.maven.model.Model;
+import org.apache.maven.model.Plugin;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.maven.model.PluginExecution;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.technologybrewery.baton.util.pom.PomHelper;
+import org.technologybrewery.baton.util.pom.PomModifications;
+
+import java.io.File;
+
+
+/**
+ * Migration that updates the POM files for machine learning pipelines with a train step to use Habushu's containerize-dependencies goal.
+ */
+public class MlTrainPipelineDockerMigration extends AbstractPomMigration {
+    protected static final Logger logger = LoggerFactory.getLogger(MlTrainPipelineDockerMigration.class);
+    private static final String DOCKER_BUILD_TYPE = "docker-build";
+    private static final String DOCKERIZE = "dockerize";
+    private static final String HABUSHU_GROUP_ID = "org.technologybrewery.habushu";
+    private static final String HABUSHU_ARTIFACT_ID = "habushu-maven-plugin";
+    private static final String FERMENTER_MDA_PLUGIN_ID = "org.technologybrewery.fermenter:fermenter-mda";
+    private static final String AISSEMBLE_TRAINING_DOCKER = "aissemble-training-docker";
+    private static final String CONTAINERIZE_DEPS_GOAL = "containerize-dependencies";
+
+    @Override
+    protected boolean shouldExecuteOnFile(File file) {
+        // package/type needs to be docker-build
+        MavenProject mavenProject = getMavenProject();
+        if (!StringUtils.equals(DOCKER_BUILD_TYPE, mavenProject.getModel().getPackaging())) {
+            return false;
+        }
+
+        return isMlTrainingDockerPom(mavenProject) && !containsHabushuContainerizeGoal(mavenProject);
+    }
+
+    @Override
+    protected boolean performMigration(File file) {
+        Model model = PomHelper.getLocationAnnotatedModel(file);
+        detectAndSetIndent(file);
+        PomModifications modifications = new PomModifications();
+
+        if (!model.getBuild().getPlugins().isEmpty()){
+            String content = getHabushuPluginConfiguration();
+            modifications.add(new PomModifications.Insertion(model.getBuild().getLocation("plugins" + END), 3, ignore -> content));
+        }
+
+        PomHelper.writeModifications(file, modifications.finalizeMods());
```

Comment:
s: PomHelper can fail and return without exception. To prevent marking the migration as successful we can:
```suggestion
        return PomHelper.writeModifications(file, modifications.finalizeMods());
```

Reply:
updated to return `PomHelper.writeModifications()`